### DISCUSSION TITLE
OG Image generation from local image not working

### DISCUSSION THREAD

**Discussion by therawk** [vercel/next.js#57335] [Help] [answered]
### Summary

I tried to use [this example](https://vercel.com/docs/functions/edge-functions/og-image-generation/og-image-examples#using-a-local-image) to generate an og image from a local image.

The documentation about this isn't really straight forward so I tried different approcahes:

I put the image in the `public` folder, I put it in the `app/og` folder and then used this code as written in the documentation:

```
const imageData = await fetch(new URL('./image.png', import.meta.url)).then(
    (res) => res.arrayBuffer(),
  );
 
  return new ImageResponse(
    (
      <div
        style={{
          display: 'flex',
          background: '#f6f6f6',
          width: '100%',
          height: '100%',
          flexDirection: 'column',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <img width="256" height="256" src={imageData} />
      </div>
    ),
    {
      width: 1200,
      height: 630,
    },
  );
```

I'm using typescript and get this error:
```
Type 'ArrayBuffer' is not assignable to type 'string'.ts(2322)
index.d.ts(2340, 9): The expected type comes from property 'src' which is declared here on type 'DetailedHTMLProps<ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement>'
```
also casting it as string doesn't work. No error is thrown but the image isn't being generated. I checked the meta section of the page and it's not there.

What am I missing here?

**Comment by mass2527** [comment]
<img width="246" alt="스크린샷 2023-10-25 오전 12 05 40" src="https://github.com/vercel/next.js/assets/70563791/2c208af1-939a-4f90-a413-6c9f7eecbb35">

## What I Know
1. Place your Open Graph image creation logic inside `app/api/og/route.tsx` instead of `app/og/route.tsx`.
2. Put your image next to the `route.tsx` file.
3. Access `http://localhost:3000/api/og`, and you can view the Open Graph image.

## What I Don't Know
I'm unsure why an ArrayBuffer type can be assigned to the src attribute, which is expected to be a string, for the img element. I attempted to investigate this but was unsuccessful, so I resorted to using @ts-ignore above the img element. Does anyone have more information on this?

**Reply by therawk** [reply]
Well yeah. I can access it via localhost but when I deploy it, it doesn't gets added to the head.
I inspected the page but nothing is there. Shouldn't it add it to the head itself?

**Reply by mass2527** [reply, marked as answer]
By utilizing [static metadata](https://nextjs.org/docs/app/building-your-application/optimizing/metadata#static-metadata) in your `page.tsx` or `layout.tsx`, og:image will be added to your head

```typescript
const URL = 'YOUR_URL'

export const metadata: Metadata = {
  title: "...",
  description: "...",
  openGraph: {
    images: [
      {
        url: `${URL}/api/og`,
        width: 256,
        height: 256,
        type: "image/png",
      },
    ],
  },
};
```
